pkg/utils/nodeclaim/nodeclaim.go

```diff
@@ -232,3 +205,39 @@ func UpdateNodeOwnerReferences(nodeClaim *v1beta1.NodeClaim, node *v1.Node) *v1.
 	})
 	return node
 }
+
+// TerminationHandler is a helper function that takes a v1beta1.NodeClaim and calls cloudProvider.Delete() if status condition
+// on nodeClaim is not terminating. If it is terminating then it will check if the kubelet process is still running. It will
+// return an error and a boolean that indicates if the reconciliation should be re-enqueued.
+func TerminationHandler(ctx context.Context, c client.Client, nodeClaim *v1beta1.NodeClaim, node *v1.Node, cloudProvider cloudprovider.CloudProvider) (bool, error) {
+	// Check if the status condition on nodeClaim is Terminating
+	terminationStatus := nodeClaim.StatusConditions().GetCondition(v1beta1.Terminating)
+	if terminationStatus.IsUnknown() || terminationStatus.IsFalse() {
+		// If not then call Delete on cloudProvider to trigger termination and always requeue reconciliation
+		if err := cloudProvider.Delete(ctx, nodeClaim); cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
+			return true, fmt.Errorf("terminating cloudprovider instance, %w", err)
```

Comment:
This function is trying to do more than just indicating if the instance is terminated. The boolean argument indicates if we want to retry reconciliation. I had it as false here when @engedaam pointed that if the termination status is not set then no matter what, we want to retry and hence I changed it to true.

Reply:
IMO, since the abstraction is: "Ensure this instance is terminated", if I'm thinking about what the function means at a high-level (ignoring implementation), I would think, "this functions should tell me if it actually _is_ terminated, and maybe fire an error if there is a transient issue". Based on that, that's what I'd have as return arguments and then let the caller figure out what to do with those values.